### PR TITLE
Adding source code of ifaddrs to be available to Android

### PR DESCRIPTION
**Pull request by javihernandez**
The full story is https://github.com/joyent/node/issues/7291.
The original source code of ifaddrs has been taken from https://github.com/kmackay/android-ifaddrs.

### CODE REVIEW

**Comment by saghul** [comment]
Hi! Thanks for submitting this! Can you please squash all commits into one and also update the autotools makefile? (Makefile.am)

File: src/unix/linux-core.c [hunk 1]
```diff
@@ -36,21 +36,20 @@
 #include <fcntl.h>
 #include <time.h>
 
-#ifndef __ANDROID__
-#define HAVE_IFADDRS_H 1
```
Comment (saghul) [inline src/unix/linux-core.c:40]: Leave the define, we might add support for a platform which doesn't have it, I guess.

File: src/unix/linux-core.c [hunk 2]
```diff
@@ -716,9 +715,6 @@ void uv_free_cpu_info(uv_cpu_info_t* cpu_infos, int count) {
 
 int uv_interface_addresses(uv_interface_address_t** addresses,
   int* count) {
-#ifndef HAVE_IFADDRS_H
-  return -ENOSYS;
-#else
```
Comment (saghul) [inline src/unix/linux-core.c:721]: Leave this, as you can see above if certain versions of UCLIBC are used this function is not supported.

**Comment by linusmartensson** [comment]
Awesome Javiér! Nice job fixing that up!

**Comment by javihernandez** [comment]
Thanks @linusmartensson, almost there!! :)

@saghul, thanks for your review. I have re-added the HAVE_IFADDRS_H condition as requested, but still have a couple of questions.

First one is about Makefile.am. Maybe I'm wrong, but we would like to have an ANDROID condition in configure.ac in order to include that files according to that condition from Makefile.am. The point is that I'm not sure how I should determine if we're building for Android by using $host_os, how should I proceed? 

The second one is about the failing tests, was it me? Isn't the new code only happening when building for Android?  It's curious that after merging master into my branch I got another extra fail. --- As a brief resume, how can I make these tests bite the dust?

Thanks again!

**Comment by saghul** [comment]
Thanks for your work!

> First one is about Makefile.am. Maybe I'm wrong, but we would like to have an ANDROID condition in configure.ac in order to include that files according to that condition from Makefile.am. The point is that I'm not sure how I should determine if we're building for Android by using $host_os, how should I proceed?

Have a look at configure.ac, I believe just by adding a AM_CONDITIONAL like this:

```
AM_CONDITIONAL([ANDROID],     [AS_CASE([$host_os], [android*],     [true], [false])])
```

Should be enough. The in Makefile.am you can check and add the files as we do with other operating systems.

> The second one is about the failing tests, was it me? Isn't the new code only happening when building for Android? It's curious that after merging master into my branch I got another extra fail. --- As a brief resume, how can I make these tests bite the dust?

We have some failing tests in some platforms, by looking at the output it doesn't look like you introduces any failures :-)

File: src/unix/linux-core.c [hunk 3]
```diff
@@ -715,93 +718,93 @@ void uv_free_cpu_info(uv_cpu_info_t* cpu_infos, int count) {
...
-  struct sockaddr_ll *sll;
-
-  if (getifaddrs(&addrs))
-    return -errno;
-
-  *count = 0;
+  #ifndef HAVE_IFADDRS_H
+    return -ENOSYS;
+  #else
+    int* count) {
```
Comment (saghul) [inline src/unix/linux-core.c]: Unintended change? This function doesn't actually need to be modified and it's actually broken as is :-)

**Comment by javihernandez** [comment]
Ok

re configure.ac; yes, it was me putting the wrong value in _--host_ when running _configure_ - I was getting nothing in $host_os :/
After having fixed that, I'll wait for your _ok_ to squash all the commits into a single one.

Thanks for all your help :)

**Comment by saghul** [comment]
Great work! LGTM. Once you squash the commits this is ready to land.

**Comment by saghul** [comment]
Thanks Javier! Landed in https://github.com/joyent/libuv/commit/a2506c936296006abe7b8a7517c3f37691191858.

**Comment by javihernandez** [comment]
Thanks Saúl,

closing report in node as well.

Cheers!

**Pull request closed without merging**